pkg/controller/cloud/node_controller.go

```diff
@@ -204,8 +211,16 @@ func (cnc *CloudNodeController) UpdateCloudNode(_, newObj interface{}) {
 		return
 	}
 
-	cloudTaint := getCloudTaint(node.Spec.Taints)
-	if cloudTaint == nil {
+	// Check if a node has been added to the initialized nodes map,
+	// meaning that it has already reached the end of the initializeNode method.
+	// If the name is found but the machine ID does not match then it is a new
+	// node using the same name as a node that was previously deleted.
+	cnc.initializedNodesMutex.Lock()
+	machineID, found := cnc.initializedNodes[node.Name]
+	cnc.initializedNodesMutex.Unlock()
+	alreadyInitialized := found && (machineID == node.Status.NodeInfo.MachineID)
```

Comment:
we might want to rely on providerID of the instance here cc @andrewsykim

Reply:
The provider ID would not be set yet I believe

Reply:
With the current state of the PR, I think it's possible for the provider ID to be not set 🤔

Reply:
on the kubelet we have: 
```golang
		if node.Spec.ProviderID == "" {
			node.Spec.ProviderID, err = cloudprovider.GetInstanceProviderID(context.TODO(), kl.cloud, kl.nodeName)
			if err != nil {
				return nil, err
			}
		}
```

Basically it relies on InstanceID call from instances interface, so as long as the cloud-provider implements this it should be fine. thoughts ?

Reply:
IIRC, that code path is only for in-tree providers. This controller only runs out-of-tree.

Reply:
True, there's also some other ways:

- store on the map the `providerID` after initialization and use `cloudprovider.GetInstanceProviderID(context.TODO(), cnc.cloud, types.NodeName(curNode.Name))` for comparison (this might be expensive as need to make call to retrieve `instanceID()`)

- change the data structure to a set of `nodeName` and make sure we add the nodeName to the set at the of the initialization, when receiving an update we can check for presence on the set

Reply:
I have just added some commits to change it to only track the node name in a set, and then remove the name from the set when it is safe to use the old behaviour of just checking the cloud taint on the event node object. This works in my tests.
For now I left in some logging that I want to remove before merging. The example output when grepping the logs for a specific node looks like
```
I0321 14:31:36.076005       1 node_controller.go:355] Successfully initialized node scaler-49-n7fhsowneckq-minion-49 with cloud provider, time=1.651611
I0321 14:31:36.076041       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:31:36.076049       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:31:36.076057       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:31:36.076062       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:01.424251       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:32:01.424255       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:01.424260       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:32:01.424265       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:31.030561       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:32:31.030565       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:32.694513       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:32:32.694520       1 node_controller.go:230] Node scaler-49-n7fhsowneckq-minion-49 is being removed from the recently initialized set as its update events no longer have the cloud taint
I0321 14:32:32.694525       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:32.694551       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:51.330381       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:33:36.002248       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
```
Once the name is removed from the set it is no problem if the node is removed on the cloud side and a new node with the same name is added.

Thoughts on this solution?

Reply:
Though without another property (machine / provider ID) to tell the difference between two nodes with the same name, there is the possibility that a node which is removed on the cloud side immediately after it is initialized in kubernetes (and therefore never sending the update event which would remove the name from the set) would not be removed from the set.
The Add event of the next node to be added with that name would still be processed, and the state would be back to normal after that, but if the Add fails and it is supposed to rely on the Update as a backup then that would not work.

Reply:
If the add fail we would still remove the nodeName from set no ?

Reply:
If the add fails there's no reason that we keep its nodeName on the set, so when receiving an update event we'd still initialize the node

Reply:
The removal of names in the set is done when the Update events see the transition of cloud taint -> no cloud taint on the event node object. If the Add event initialization for the second node with the same name failed and it wasn't able to do the initialization on an Update event because the name already exists in the set, it would never be initialized so that transition would never be seen.

Maybe going back to a map and having the value in the map be the time that the name is added would be a way to check if a name should really still be tracked or not.

Reply:
@tghartland - if we are not able to initialize the node we shouldn't add it to the set, so if there's an error on the add we won't add the nodeName.
If we remove the node right after initialization and add it back, the add event will be processed and start initialization, with 2 possible outcomes I think:

- initialization is achieved, we add the nodeName from the set
- initialization failed, we remove the nodeName from the set if it exists

when receiving an update after the add event the set will behave as it should since if the initialization is done the nodeName would be on the set.

Reply:
> initialization failed, we remove the nodeName from the set if it exists

In fact it doesn't even have to fail, if the name is in the set at the start of an Add event then it must have been left over from a previous node with the same name, and should be removed.
Commit fixing this: https://github.com/kubernetes/kubernetes/pull/75405/commits/7f12d7dcf75e74e19f601b9f9518165937b6e3c9
I think this takes care of any issues to do with a name not being removed properly.

Reply:
This would work too. more generally, seeing an add should clean the set from the nodeName